### DISCUSSION TITLE
How to override a built-in extension in jupyter-book?

### DISCUSSION THREAD

**Discussion by akhmerov** [executablebooks/meta#459] [Q&A]
I'm playing around with enabling sphinx-design (I know it's super early :upside_down_face:), and I observe that

- it conflicts with sphinx-panels (sphinx-panels overrides many of design directives and nodes)
- it seems impossible to override sphinx-panels

Hence my question: how to do so?

**Comment by chrisjsewell** [comment]
Well you can’t at present (disable sphinx-panels) 😬 
I’ve checked that sphinx-design does now work with sphinx-panels (if added after sphinx_panels in the extension list), except it emits  warning messages about the directive/node overrides.
I think maybe you can suppress the requisite warnings: https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-suppress_warnings?

**Reply by chrisjsewell** [reply]
Or are you saying that sphinx panels is after sphinx design in the extension list?

**Reply by akhmerov** [reply]
I think so: adding sphinx design to extra extensions in jupyter-book config leaves me with a dropdown from sphinx panels

**Reply by akhmerov** [reply]
Hm, I wasn't exactly right.

My book config (the relevant part):
```yaml
  extra_extensions:
  - sphinx_design
```
I also confirmed that `sphinx_design` is included in the extension list *after* than the `sphinx_panels` by inspecting the sphinx app.

Myst source
```
:::{dropdown} Child dropdown title
:color: warning
:icon: alert

Dropdown content
:::
```

Warnings I see:

```
WARNING: while setting up extension sphinx_design: node class 'dropdown_main' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx_design: node class 'dropdown_title' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx_design: directive 'dropdown' is already registered, it will be overridden
WARNING: while setting up extension sphinx_design: role 'fa' is already registered, it will be overridden
WARNING: while setting up extension sphinx_design: node class 'fontawesome' is already registered, its visitors will be overridden
```
So far this is exactly like you describe, and like I'd expect.

**However**, this produces the HTML
```html
<details class="sphinx-bs dropdown card sd-mb-3">
<summary class="summary-title card-header sd-bg-warning sd-bg-text-warning">
Child dropdown title<div class="summary-down docutils">
<svg version="1.1" width="24" height="24" class="octicon octicon-chevron-down" viewBox="0 0 24 24" aria-hidden="true"><path fill-rule="evenodd" d="M5.22 8.72a.75.75 0 000 1.06l6.25 6.25a.75.75 0 001.06 0l6.25-6.25a.75.75 0 00-1.06-1.06L12 14.44 6.28 8.72a.75.75 0 00-1.06 0z"></path></svg></div>
<div class="summary-up docutils">
<svg version="1.1" width="24" height="24" class="octicon octicon-chevron-up" viewBox="0 0 24 24" aria-hidden="true"><path fill-rule="evenodd" d="M18.78 15.28a.75.75 0 000-1.06l-6.25-6.25a.75.75 0 00-1.06 0l-6.25 6.25a.75.75 0 101.06 1.06L12 9.56l5.72 5.72a.75.75 0 001.06 0z"></path></svg></div>
</summary><div class="summary-content card-body docutils">
<p class="card-text">Dropdown content</p>
</div>
</details>
```
On the other hand, if I comment out `sphinx-panels` in the source of jupyter_book/config.py, then I get the following HTML:
```html
<details class="sd-sphinx-override sd-dropdown sd-card sd-mb-3">
<summary class="sd-summary-title sd-card-header sd-bg-warning sd-bg-text-warning">
<span class="sd-summary-icon"><svg version="1.1" width="1.0em" height="1.0em" class="sd-octicon sd-octicon-alert" viewBox="0 0 16 16" aria-hidden="true"><path fill-rule="evenodd" d="M8.22 1.754a.25.25 0 00-.44 0L1.698 13.132a.25.25 0 00.22.368h12.164a.25.25 0 00.22-.368L8.22 1.754zm-1.763-.707c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0114.082 15H1.918a1.75 1.75 0 01-1.543-2.575L6.457 1.047zM9 11a1 1 0 11-2 0 1 1 0 012 0zm-.25-5.25a.75.75 0 00-1.5 0v2.5a.75.75 0 001.5 0v-2.5z"></path></svg></span>Child dropdown title<div class="sd-summary-down docutils">
<svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-down" viewBox="0 0 24 24" aria-hidden="true"><path fill-rule="evenodd" d="M5.22 8.72a.75.75 0 000 1.06l6.25 6.25a.75.75 0 001.06 0l6.25-6.25a.75.75 0 00-1.06-1.06L12 14.44 6.28 8.72a.75.75 0 00-1.06 0z"></path></svg></div>
<div class="sd-summary-up docutils">
<svg version="1.1" width="1.5em" height="1.5em" class="sd-octicon sd-octicon-chevron-up" viewBox="0 0 24 24" aria-hidden="true"><path fill-rule="evenodd" d="M18.78 15.28a.75.75 0 000-1.06l-6.25-6.25a.75.75 0 00-1.06 0l-6.25 6.25a.75.75 0 101.06 1.06L12 9.56l5.72 5.72a.75.75 0 001.06 0z"></path></svg></div>
</summary><div class="sd-summary-content sd-card-body docutils"></details>
```

So after all, `sphinx_panels` does interfere with `sphinx_design`.

**Reply by chrisjsewell** [reply]
> I also confirmed that sphinx_design is included in the extension list after than the sphinx_panels by inspecting the sphinx app.

no I think it is after: you can see this by running `jupyter-book config sphinx`

**Reply by akhmerov** [reply]
Yes, that's what I wrote. It's also what I see:
```
extensions = ['sphinx_togglebutton', 'sphinx_copybutton', 'myst_nb', 'jupyter_book', 'sphinx_thebe', 'sphinx_comments', 'sphinx_external_toc', 'sphinx.ext.intersphinx', 'sphinx_panels', 'sphinx_book_theme', 'sphinxext.opengraph', 'sphinx_design', 'sphinx_jupyterbook_latex']
```

Looking at the html classes you see that *some* are indeed from sphinx-design, while others are presumably from sphinx-
panels, resulting in a weird mixture.

**Reply by chrisjsewell** [reply]
Oh it’s maybe because of the post-transforms, this should only be affecting dropdown

**Reply by chrisjsewell** [reply]
Should just need to increase the relative priority of the sphinx-design one

**Reply by akhmerov** [reply]
It's too bad that I'd like to use the dropdown :upside_down_face: 

However `jupyter-book config sphinx` seems quite handy.

![image](https://user-images.githubusercontent.com/2069677/129339957-f2ce46dd-c589-4c9f-be6c-9fc0ce350563.png)

**Reply by chrisjsewell** [reply]
https://github.com/executablebooks/jupyter-book/pull/1422
